R/http_methods.R

```diff
@@ -0,0 +1,185 @@
+  if(isTRUE(local)) {
+    return(read_local_protocol())
+  }
+  fetch_json(host, port, secure, "protocol")
+}
+
+#' @rdname http-methods
+#' @export
+new_tab <- function(host = "localhost", port = 9222, secure = FALSE, url = NULL) {
+  if(!is.null(url)) {
+    assert_that(is_scalar_character(url))
+  }
+  fetch_json(host, port, secure, "new", url)
+}
+
+#' @rdname http-methods
+#' @export
+activate_target <- function(
+  host = "localhost", port = 9222, secure = FALSE, target_id
+) {
+  target_method(host, port, secure, target_id, "activate")
+}
+
+#' @rdname http-methods
+#' @export
+close_target <- function(
+  host = "localhost", port = 9222, secure = FALSE, target_id
+) {
+  target_method(host, port, secure, target_id, "close")
+}
+
+#' @rdname http-methods
+#' @export
+inspect_target <- function(
+  host = "localhost", port = 9222, secure = FALSE, target_id
+) {
+  assert_that(is_scalar_character(target_id))
+  targets <- list_targets(host, port, secure)
+  ids <- purrr::map_chr(targets, "id")
+  index <- purrr::detect_index(ids, ~ identical(.x, target_id))
+  if(index == 0) {
+    stop("Wrong target_id.")
+  }
+  if(!interactive()) {
+    warning("The inspect method can only be used in an interactive session.")
+    return(NULL)
+  }
+  path <- purrr::pluck(targets, index, "devtoolsFrontendUrl")
+  url <- build_http_url(host, port, secure, path)
+  browse_url(url)
+}
+
+fetch_json <- function(host, port, secure, method, query = NULL) {
+  check_host_port_args(host, port)
+  url <- build_http_url(host, port, secure, path = c("json", method), query)
+  from_json(url)
+}
+
+target_method <- function(host, port, secure, target_id, method) {
+  check_host_port_args(host, port)
+  assert_that(is_scalar_character(target_id))
+  url <- build_http_url(host, port, secure, path = c("json", method, target_id))
+  res <- httr::GET(url, httr::use_proxy(""))
+  !httr::http_error(res)
+}
+
+check_host_port_args <- function(host, port) {
+  assert_that(is_scalar_character(host))
+  assert_that(is.scalar(port))
+  if(!is.number(port) && !is_scalar_character(port)) {
+    stop("port must be a character or a numeric scalar.")
+  }
+}
+
+from_json <- function(path) {
+  jsonlite::fromJSON(
+    path,
+    simplifyVector = TRUE,
+    simplifyDataFrame = FALSE,
+    simplifyMatrix = FALSE
+  )
+}
+
+read_local_protocol <- function() {
+  js_protocol <- from_json(local_protocol_file("js"))
+  browser_protocol <- from_json(local_protocol_file("browser"))
+  browser_protocol$domains <- c(browser_protocol$domains, js_protocol$domains)
+  browser_protocol
+}
+
+local_protocol_file <- function(file = c("js", "browser")) {
+  file <- match.arg(file)
+  system.file("protocol", paste0(file, "_protocol.json"), package = "crrri", mustWork = TRUE)
+}
+
+browse_url <- function(url) {
+  localhost <- is_localhost(url)
+  viewer <- getOption("viewer")
+  if (is.null(viewer) || !localhost){
+    utils::browseURL(url)
+  }
+  else {
+    # we know here that we are in rstudio
```

Comment:
I only set the `rstudioapi` in `Suggests` not in `Imports`. I modified this call in https://github.com/RLesur/crrri/pull/56/commits/307c50f4e8865d33e3f4580726a0d2104c25be4d and use `requireNamespace()` first

Reply:
This means that if someone doesn't have `rstudioapi` 📦, it will not be able to inspect in RStudio. `rstudioapi` does not come with RStudio I think - we may have question in the future about why inspect does not work in RStudio. 
I am not sure we can reproduce the `getVersion` easily, but I think we can make a test less strict.
We'll see on the run with people using it.